File: ui/src/lib/attachments.ts
// Composer-side attachment handling (TD-1709).
//
// This is the *courtesy* half of the gate. The daemon refuses an oversize or
// binary attachment on arrival no matter what a client sends — see
// core/tstd/attachments.py — and nothing here weakens or replaces that. What
// it buys is copy: a file refused at the composer names itself the moment the
// user drops it, instead of after a round trip that also swallowed the
// message they typed alongside it.
//
// The two sides run the same test on purpose: strict UTF-8 plus a NUL scan,
// against the same caps. The caps arrive on `boundary_update` rather than
// being hardcoded here (§6: the UI never derives truth it wasn't given).
//
// Pure and DOM-free apart from TextDecoder (available in node and the
// browser), so it unit-tests under vitest's node environment.

import type { Attachment, AttachmentLimits } from "./protocol";

/** Mirrors core/tstd/attachments.py. Used only until the first
 *  `boundary_update` names the workspace's real numbers. */
export const DEFAULT_ATTACHMENT_LIMITS: AttachmentLimits = {
  max_file_bytes: 256_000,
  max_total_bytes: 512_000,
  max_count: 10,
};

/** An accepted file, before the composer gives it a row id. */
export interface NewAttachment {
  name: string;
  /** Byte length of the file, which is what the caps are measured against. */
  size: number;
  /** The file's bytes, base64'd for the wire. */
  content_b64: string;
}

/** One accepted file, held in the composer until the message goes out. */
export interface AttachmentDraft extends NewAttachment {
  /** Composer-local id; never a conversation message id. */
  id: string;
}

/** A file the composer would not take, and the sentence explaining why. */
export interface AttachmentRefusal {
  name: string;
  /** Matches the daemon's wire error code for the same condition. */
  code: string;
  message: string;
}

export type AttachmentOutcome =
  | { ok: true; draft: NewAttachment }
  | { ok: false; refusal: AttachmentRefusal };

/** Sizes for humans — the same shape the daemon's refusals use. */
export function formatBytes(count: number): string {
  if (count < 1000) return `${count} bytes`;
  if (count < 1_000_000) return `${(count / 1000).toFixed(1)} KB`;
  return `${(count / 1_000_000).toFixed(1)} MB`;
}

/** Bytes already staged, which the per-message total is measured against. */
export function totalBytes(drafts: readonly NewAttachment[]): number {
  return drafts.reduce((sum, d) => sum + d.size, 0);
}

/** Basename only. The daemon strips this again — it does not trust us — but
 *  a chip reading `passwd` rather than `../../etc/passwd` is the honest
 *  preview of what will actually be sent. */
export function safeName(raw: string): string {
  return raw.replace(/\\/g, "/").split("/").pop()?.trim() ?? "";
}

/** Strict UTF-8 plus a NUL scan: the whole text/binary test, same as the
 *  daemon's. `fatal` is what makes it a test rather than a lossy decode —
 *  without it a PNG comes back as replacement characters and reads as prose. */
export function isImageBytes(bytes: Uint8Array): boolean {
  if (
    bytes.length >= 8 &&
    bytes[0] === 0x89 &&
    bytes[1] === 0x50 &&
    bytes[2] === 0x4e &&
    bytes[3] === 0x47
  ) {
    return true;
  }
  if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
    return true;
  }
  if (bytes.length >= 6) {
    const ascii = String.fromCharCode(...bytes.subarray(0, 6));
    if (ascii === "GIF87a" || ascii === "GIF89a") return true;
  }
  if (
    bytes.length >= 12 &&
    String.fromCharCode(...bytes.subarray(0, 4)) === "RIFF" &&
    String.fromCharCode(...bytes.subarray(8, 12)) === "WEBP"
  ) {
    return true;
  }
  return false;
}

export function isTextBytes(bytes: Uint8Array): boolean {
  if (bytes.includes(0)) return false;
  try {
    new TextDecoder("utf-8", { fatal: true }).decode(bytes);
    return true;
  } catch {
    return false;
  }
}

/** Base64 without a data-URL round trip; chunked so a large file cannot blow
 *  the argument limit on String.fromCharCode. */
export function toBase64(bytes: Uint8Array): string {
  const CHUNK = 0x8000;
  let binary = "";
  for (let i = 0; i < bytes.length; i += CHUNK) {
    binary += String.fromCharCode(...bytes.subarray(i, i + CHUNK));
  }
  return btoa(binary);
}

/** Magic-byte image detection — same set as core/tstd/attachments.py (TD-4705). */
export function detectImageMime(bytes: Uint8Array): string | null {
  if (
    bytes.length >= 8 &&
    bytes[0] === 0x89 &&
    bytes[1] === 0x50 &&
    bytes[2] === 0x4e &&
    bytes[3] === 0x47 &&
    bytes[4] === 0x0d &&
    bytes[5] === 0x0a &&
    bytes[6] === 0x1a &&
    bytes[7] === 0x0a
  ) {
    return "image/png";
  }
  if (bytes.length >= 3 && bytes[0] === 0xff && bytes[1] === 0xd8 && bytes[2] === 0xff) {
    return "image/jpeg";
  }
  if (
    bytes.length >= 6 &&
    ((bytes[0] === 0x47 &&
      bytes[1] === 0x49 &&
      bytes[2] === 0x46 &&
      bytes[3] === 0x38 &&
      (bytes[4] === 0x37 || bytes[4] === 0x39) &&
      bytes[5] === 0x61) ||
      (bytes[0] === 0x47 &&
        bytes[1] === 0x49 &&
        bytes[2] === 0x46 &&
        bytes[3] === 0x38 &&
        bytes[4] === 0x39 &&
        bytes[5] === 0x61))
  ) {
    return "image/gif";
  }
  if (
    bytes.length >= 12 &&
    bytes[0] === 0x52 &&
    bytes[1] === 0x49 &&
    bytes[2] === 0x46 &&
    bytes[3] === 0x46 &&
    bytes[8] === 0x57 &&
    bytes[9] === 0x45 &&
    bytes[10] === 0x42 &&
    bytes[11] === 0x50
  ) {
    return "image/webp";
  }
  return null;
}

/** Vet one dropped/picked/pasted file against the caps and the text test.
 *
 *  `staged` is what the composer already holds, so the count and the running
 *  total are judged against the message as it will actually be sent. Every
 *  refusal names the file and the way forward, and none of them claims
 *  anything was sent — nothing has been.
 *
 *  When `allowImages` is true (from `tier_state.vision`, TD-4705), PNG/JPEG/
 *  GIF/WebP pass; other binary still refuses. When false, images refuse with
 *  `attachment_no_vision` rather than the generic binary copy. */
export function acceptAttachment(
  rawName: string,
  bytes: Uint8Array,
  limits: AttachmentLimits,
  staged: readonly NewAttachment[] = [],
  allowImages = false,
): AttachmentOutcome {
  const name = safeName(rawName) || "file";

  if (staged.length >= limits.max_count) {
    return {
      ok: false,
      refusal: {
        name,
        code: "attachment_too_many",
        message: `You can attach ${limits.max_count} file${limits.max_count === 1 ? "" : "s"} to one message. Send these first, or raise max_count under attachments: in .tst/config.yaml.`,
      },
    };
  }

  if (bytes.length > limits.max_file_bytes) {
    return {
      ok: false,
      refusal: {
        name,
        code: "attachment_too_large",
        message: `${name} is ${formatBytes(bytes.length)}, and the limit is ${formatBytes(limits.max_file_bytes)} per file. Attach a smaller file, or raise max_file_bytes under attachments: in .tst/config.yaml.`,
      },
    };
  }

  if (totalBytes(staged) + bytes.length > limits.max_total_bytes) {
    return {
      ok: false,
      refusal: {
        name,
        code: "attachment_total_too_large",
        message: `Adding ${name} would put this message over ${formatBytes(limits.max_total_bytes)} of attachments. Send what's here first, or raise max_total_bytes under attachments: in .tst/config.yaml.`,
      },
    };
  }

  const mime = detectImageMime(bytes);
  if (mime !== null) {
    if (!allowImages) {
      return {
        ok: false,
        refusal: {
          name,
          code: "attachment_no_vision",
          message: `${name} is an image, but the active model does not accept images. Set vision: true on that tier in config.yaml, or switch to a vision-capable preset.`,
        },
      };
    }
    return {
      ok: true,
      draft: { name, size: bytes.length, content_b64: toBase64(bytes) },
    };
  }

  if (!isTextBytes(bytes)) {
    return {
      ok: false,
      refusal: {
        name,
        code: "attachment_binary",
        message: `${name} isn't a text file or a PNG/JPEG/GIF/WebP image, so it can't be attached.`,
      },
    };
  }

  return {
    ok: true,
    draft: { name, size: bytes.length, content_b64: toBase64(bytes) },
  };
}

/** The wire shape, stripped of the composer-local bookkeeping. */
export function toWireAttachments(drafts: readonly NewAttachment[]): Attachment[] {
  return drafts.map((d) => ({ name: d.name, content_b64: d.content_b64 }));
}

/** What a sent row shows: the chip's label, without the bytes riding along in
 *  the transcript for the life of the session. */
export interface AttachmentChip {
  name: string;
  size: number;
}

export function toChips(drafts: readonly NewAttachment[]): AttachmentChip[] {
  return drafts.map((d) => ({ name: d.name, size: d.size }));
}

File: core/tstd/attachments.py
"""Text-file and image attachments on a user message (TD-1709, TD-4705).

**The daemon is the gate, not the composer.**  A client is not trustworthy:
it can be an older build, a stalled one, or something that is not our UI at
all.  So the caps and the text/binary test live here, on the receiving side,
and the composer's early refusal is a courtesy that buys better copy — never
the thing that actually holds the wall.

Attachments arrive as base64 so the daemon sees the *bytes*, not a decode a
client already made on our behalf.  That is what makes "is this text?"
answerable at all: a client that reads a PNG and ships it raw fails the
strict UTF-8 decode here, where a pre-decoded string would have arrived as
replacement characters and looked like ordinary prose.

Deliberately leaf: this module imports nothing from ``tstd``, so both
``protocol`` (the wire) and ``boundary_config`` (the caps) can depend on it
without a cycle.
"""

from __future__ import annotations

import base64
import binascii
from collections.abc import Sequence
from dataclasses import dataclass
from typing import Any, Protocol

from pydantic import BaseModel, Field

# Defaults when `.tst/config.yaml` declares no `attachments:` section.  Sized
# against the websocket frame ceiling (1 MiB, the `websockets` default): the
# total cap plus base64's third again has to fit inside one frame, or an
# oversize message would be dropped by the transport before this module ever
# got the chance to refuse it with copy the user can act on.
DEFAULT_MAX_FILE_BYTES = 256_000
DEFAULT_MAX_TOTAL_BYTES = 512_000
DEFAULT_MAX_COUNT = 10


class AttachmentLimits(BaseModel):
    """Caps on what one message may carry (`.tst/config.yaml` ``attachments``).

    Also ridden out on ``boundary_update`` so the composer refuses against
    the workspace's real numbers rather than a hardcoded guess (§6: the UI
    never derives truth it wasn't given).
    """

    max_file_bytes: int = Field(default=DEFAULT_MAX_FILE_BYTES, ge=1)
    max_total_bytes: int = Field(default=DEFAULT_MAX_TOTAL_BYTES, ge=1)
    max_count: int = Field(default=DEFAULT_MAX_COUNT, ge=1)


class AttachmentError(Exception):
    """A refused attachment, carrying the wire error code and its copy."""

    def __init__(self, code: str, message: str) -> None:
        self.code = code
        self.message = message
        super().__init__(f"{code}: {message}")


class AttachmentLike(Protocol):
    """The shape this module needs off the wire message.

    Structural rather than a concrete import: ``protocol.Attachment`` is the
    mirrored wire definition and must stay readable in ``protocol.py``, which
    already imports from here for the limits.
    """

    @property
    def name(self) -> str: ...

    @property
    def content_b64(self) -> str: ...


@dataclass(frozen=True)
class DecodedAttachment:
    """One attachment the daemon has accepted, decoded and measured itself.

    Images carry ``data`` and ``media_type`` and leave ``text`` empty so a
    native loop can mention the file without dumping bytes into the prompt.
    """

    name: str
    size: int
    text: str | None = None
    mime: str | None = None
    data_b64: str | None = None
    data: bytes | None = None
    media_type: str | None = None

    @property
    def is_image(self) -> bool:
        return self.mime is not None or self.media_type is not None


def format_bytes(count: int) -> str:
    """Sizes for humans, in the copy a refusal shows."""
    if count < 1000:
        return f"{count} bytes"
    if count < 1_000_000:
        return f"{count / 1000:.1f} KB"
    return f"{count / 1_000_000:.1f} MB"


def _safe_name(raw: str) -> str:
    """The basename, with anything that could forge a delimiter refused."""
    name = raw.replace("\\", "/").rsplit("/", 1)[-1].strip()
    if name in ("", ".", ".."):
        raise AttachmentError(
            "attachment_bad_name",
            f"An attachment arrived with an unusable file name ({raw!r}). "
            "Re-attach the file. Nothing was sent.",
        )
    if any(ch < " " or ch == "\x7f" for ch in name):
        raise AttachmentError(
            "attachment_bad_name",
            f"The file name {name!r} contains control characters and was refused. "
            "Rename the file and re-attach it. Nothing was sent.",
        )
    return name


def detect_image_mime(raw: bytes) -> str | None:
    """Magic-byte image detection. No resize — bytes are sent as-is (TD-4705)."""
    if raw.startswith(b"\x89PNG\r\n\x1a\n"):
        return "image/png"
    if raw.startswith(b"\xff\xd8\xff"):
        return "image/jpeg"
    if raw.startswith((b"GIF87a", b"GIF89a")):
        return "image/gif"
    if len(raw) >= 12 and raw[:4] == b"RIFF" and raw[8:12] == b"WEBP":
        return "image/webp"
    return None


def sniff_image(raw: bytes) -> str | None:
    """Return a media type for a well-known image header, else None."""
    return detect_image_mime(raw)


def _decode_text(name: str, raw: bytes) -> str:
    """Strict UTF-8, plus a NUL scan — the whole text/binary test."""
    binary = AttachmentError(
        "attachment_binary",
        f"{name} isn't a text file or a PNG/JPEG/GIF/WebP image, so it can't "
        "be attached. Attach a text file, or enable vision on the active "
        "tier in config.yaml if this is an image. Nothing was sent.",
    )
    if b"\x00" in raw:
        raise binary
    try:
        return raw.decode("utf-8")
    except UnicodeDecodeError as e:
        raise binary from e


def _no_vision_refusal(name: str) -> AttachmentError:
    return AttachmentError(
        "attachment_no_vision",
        f"{name} is an image, but the active model does not accept images. "
        "Set vision: true on that tier in config.yaml, or switch to a "
        "vision-capable preset. Nothing was sent.",
    )


def decode_attachments(
    items: Sequence[AttachmentLike],
    limits: AttachmentLimits,
    *,
    allow_images: bool = False,
) -> list[DecodedAttachment]:
    """Decode and vet every attachment on a message, or refuse the message."""
    if not items:
        return []

    if len(items) > limits.max_count:
        raise AttachmentError(
            "attachment_too_many",
            f"{len(items)} files attached; this workspace allows "
            f"{limits.max_count} per message. Send fewer, or raise max_count "
            "under attachments: in .tst/config.yaml. Nothing was sent.",
        )

    decoded: list[DecodedAttachment] = []
    total = 0
    for item in items:
        name = _safe_name(item.name)
        try:
            raw = base64.b64decode(item.content_b64, validate=True)
        except (binascii.Error, ValueError) as e:
            raise AttachmentError(
                "attachment_invalid_encoding",
                f"{name} arrived in a form the daemon couldn't read. "
                "Re-attach it. Nothing was sent.",
            ) from e

        if len(raw) > limits.max_file_bytes:
            raise AttachmentError(
                "attachment_too_large",
                f"{name} is {format_bytes(len(raw))}; this workspace allows "
                f"{format_bytes(limits.max_file_bytes)} per file. Attach a "
                "smaller file, or raise max_file_bytes under attachments: in "
                ".tst/config.yaml. Nothing was sent.",
            )

        total += len(raw)
        if total > limits.max_total_bytes:
            raise AttachmentError(
                "attachment_total_too_large",
                f"The attachments total more than "
                f"{format_bytes(limits.max_total_bytes)}, which is this "
                "workspace's limit for one message. Send them across separate "
                "messages, or raise max_total_bytes under attachments: in "
                ".tst/config.yaml. Nothing was sent.",
            )

        mime = detect_image_mime(raw)
        if mime is not None:
            if not allow_images:
                raise _no_vision_refusal(name)
            decoded.append(
                DecodedAttachment(
                    name=name,
                    size=len(raw),
                    text="",
                    mime=mime,
                    media_type=mime,
                    data=raw,
                    data_b64=base64.b64encode(raw).decode("ascii"),
                )
            )
            continue

        text = _decode_text(name, raw)
        decoded.append(DecodedAttachment(name=name, text=text, size=len(raw)))

    return decoded


def render_user_content(text: str, attachments: Sequence[DecodedAttachment]) -> str:
    """Fold accepted text attachments into the message the model reads."""
    text_files = [a for a in attachments if not a.is_image]
    if not text_files:
        image_note = _image_summary(attachments)
        if image_note and text:
            return f"{text}\n\n{image_note}"
        if image_note:
            return image_note
        return text

    blocks = [text] if text else []
    for item in text_files:
        assert item.text is not None
        blocks.append(
            f"--- attached file: {item.name} ({item.size} bytes) ---\n"
            f"{item.text}\n"
            f"--- end of {item.name} ---"
        )
    body = "\n\n".join(blocks)
    image_note = _image_summary(attachments)
    if image_note:
        body = f"{body}\n\n{image_note}" if body else image_note
    return body


def _image_summary(attachments: Sequence[DecodedAttachment]) -> str:
    images = [a for a in attachments if a.is_image]
    if not images:
        return ""
    names = ", ".join(f"{a.name} ({a.size} bytes)" for a in images)
    return f"[Images attached: {names}]"


def build_provider_user_content(
    text: str,
    attachments: Sequence[DecodedAttachment],
) -> str | list[dict[str, Any]]:
    """OpenAI-compatible user content — text and/or image_url parts (TD-4705)."""
    images = [a for a in attachments if a.is_image]
    if not images:
        return render_user_content(text, attachments)

    parts: list[dict[str, Any]] = []
    prose = render_user_content(text, [a for a in attachments if not a.is_image])
    if prose:
        parts.append({"type": "text", "text": prose})
    for item in images:
        assert item.mime is not None and item.data_b64 is not None
        parts.append(
            {
                "type": "image_url",
                "image_url": {"url": f"data:{item.mime};base64,{item.data_b64}"},
            }
        )
    return parts
